claims_common.go

```diff
@@ -0,0 +1,339 @@
+		v <= SecurityLifecycleSecuredMax {
+		return "secured"
+	}
+
+	if v >= SecurityLifecycleNonPsaRotDebugMin &&
+		v <= SecurityLifecycleNonPsaRotDebugMax {
+		return "non-psa-rot-debug"
+	}
+
+	if v >= SecurityLifecycleRecoverablePsaRotDebugMin &&
+		v <= SecurityLifecycleRecoverablePsaRotDebugMax {
+		return "recoverable-psa-rot-debug"
+	}
+
+	if v >= SecurityLifecycleDecommissionedMin &&
+		v <= SecurityLifecycleDecommissionedMax {
+		return "decommissioned"
+	}
+
+	return "invalid"
+}
+
+func isValidSecurityLifeCycle(v uint16) error {
+	// Accept any security lifecycle in the state machine, including values that
+	// can't produce trustable PSA evidence.
+	if securityLifeCycleToString(v) == "invalid" {
+		return fmt.Errorf("%w: value %d is invalid", ErrWrongSyntax, v)
+	}
+
+	return nil
+}
+
+func isValidImplID(v []byte) error {
+	l := len(v)
+
+	if l != ImplIDLen {
+		return fmt.Errorf(
+			"%w: invalid length %d (MUST be %d bytes)",
+			ErrWrongSyntax, l, ImplIDLen,
+		)
+	}
+
+	return nil
+}
+
+func isValidBootSeed(v []byte, profile string) error {
+	l := len(v)
+
+	switch profile {
+	case PsaProfile2:
+		if l < 8 || l > 32 {
+			return fmt.Errorf(
+				"%w: invalid length %d (MUST be between 8 and 32 bytes)",
+				ErrWrongSyntax, l,
+			)
+		}
+	case PsaProfile1:
+		if l != 32 {
+			return fmt.Errorf(
+				"%w: invalid length %d (MUST be 32 bytes)",
+				ErrWrongSyntax, l,
+			)
+		}
+	}
+
+	return nil
+}
+
+func isValidCertificationReference(v string) error {
+	notDigit := func(c rune) bool { return c < '0' || c > '9' }
+
+	if len(v) != 13 || strings.IndexFunc(v, notDigit) != -1 {
+		return fmt.Errorf(
+			"%w: MUST be in EAN-13 format",
+			ErrWrongSyntax,
+		)
+	}
+
+	return nil
+}
+
+func isPSAHashType(b []byte) error {
+	l := len(b)
+
+	if l != 32 && l != 48 && l != 64 {
+		return fmt.Errorf(
+			"%w: length %d (psa-hash-type MUST be 32, 48 or 64 bytes)",
+			ErrWrongSyntax, l,
+		)
+	}
+
+	return nil
+}
+
+func isValidInstID(v []byte) error {
+	l := len(v)
+
+	if l != InstIDLen {
+		return fmt.Errorf(
+			"%w: invalid length %d (MUST be %d bytes)",
+			ErrWrongSyntax, l, InstIDLen,
+		)
+	}
+
+	if v[0] != 0x01 {
+		return fmt.Errorf(
+			"%w: invalid EUID type (MUST be RAND=0x01)",
+			ErrWrongSyntax,
+		)
+	}
+
+	return nil
+}
+
+func isValidVSI(v string) error {
+	// https://github.com/thomas-fossati/draft-psa-token/issues/59
+	if v == "" {
+		return fmt.Errorf("%w: empty string", ErrWrongSyntax)
+	}
+
+	return nil
+}
+
+func isValidSwComponents(scs []SwComponent) error {
+	if len(scs) == 0 {
+		return fmt.Errorf("%w: there MUST be at least one entry", ErrWrongSyntax)
+	}
+
+	for i, sc := range scs {
+		if err := sc.Validate(); err != nil {
+			return fmt.Errorf("failed at index %d: %w", i, err)
+		}
+	}
+
+	return nil
+}
+
+func isValidNonce(v []byte) error {
+	if err := isPSAHashType(v); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setClientID(dst **int32, src *int32) error {
+	// any int32 value is acceptable
```

Comment:
Should we check for nil pointer here?

Reply:
No, these are all internal functions.  We trust the caller to do the right thing.

Reply:
In Sw nothing can be trusted, perhaps will help debug some memory leaks or random corruptions where if accidentally the pointer is set to null, then we continue rather than crash?

Reply:
> In Sw nothing can be trusted

uhm... I don't think you usually to compile by hand and hammer the bits directly into memory with a stick right? :-)

Reply:
> perhaps will help debug some memory leaks

go is garbage collected, but even if it weren't checking for not-nil would not catch any memory leaks

> or random corruptions where if accidentally the pointer is set to null, then we continue rather than crash?

if a caller of this will supply a nil pointer then it's an invariant violation and in such case we should `panic`.  But the go runtime does that for us already, so there is no need to duplicate functionality.